src/core/stdcpp/allocator.d

```diff
@@ -0,0 +1,126 @@
+/**
+ * D binding to C++ std::allocator.
+ *
+ * Copyright: Copyright (c) 2019 D Language Foundation
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Manu Evans
+ * Source:    $(DRUNTIMESRC core/stdcpp/allocator.d)
+ */
+
+module core.stdcpp.allocator;
+
+import core.stdcpp.new_;
+import core.stdcpp.xutility : __cpp_sized_deallocation, __cpp_aligned_new;
+
+// some versions of VS require a `* const` pointer mangling hack
+// we need a way to supply the target VS version to the compile
+version (CppRuntime_Microsoft)
+    version = NeedsMangleHack;
+
+/**
+ * Allocators are classes that define memory models to be used by some parts of
+ * the C++ Standard Library, and most specifically, by STL containers.
+ */
+extern(C++, class)
+extern(C++, "std")
+struct allocator(T)
+{
+    static assert(!is(T == const), "The C++ Standard forbids containers of const elements because allocator!(const T) is ill-formed.");
+    static assert(!is(T == immutable), "immutable is not representable in C++");
+    static assert(!is(T == class), "Instantiation with `class` is not supported; D can't mangle the base (non-pointer) type of a class. Use `extern (C++, class) struct T { ... }` instead.");
+
+    ///
+    alias value_type = T;
+
+    version (CppRuntime_Microsoft)
+    {
+        ///
+        T* allocate(size_t count) @nogc;
+        ///
+        void deallocate(T* ptr, size_t count) @nogc;
+
+        ///
+        enum size_t max_size = size_t.max / T.sizeof;
+
+        version (NeedsMangleHack)
+        {
+            // HACK: workaround to make `deallocate` link as a `T * const`
+            private extern (D) enum string constHack(string name) = (){
```

Comment:
@TurkeyMan 
This is always a pain point when trying to link to C++ libraries. I've been thinking of it and possible adding a special "cpp_const" symbol that can only be used in extern(C++) functions would be worth implementing?

```D
extern(C++) void test( cpp_const char* a, cpp_const char* cpp_const b, char* cpp_const c )
{
    pragma(msg, typeof(a)); // D type: const(char)*
    pragma(msg, typeof(b)); // D type: const(char*)
    pragma(msg, typeof(c)); // D type: char* 
                            // Note:    ^ don't actually get const here, would only be for mangling
}
```

Thoughts ?

Reply:
Yah same here, the best option would be to have const behave the same way in C++ so we can do the same mangling as C++. It'd be the simplest cleanest solution. The problem is D's const behaves differently and probably should stay the way it is. So I was thinking a new keyword could be used but it obviously won't be as nice as just using const.